src/allocation_engine/interval_tree.rs

```diff
@@ -424,6 +424,34 @@ impl InnerNode {
             },
         }
     }
+
+    /// Delete `key` from the subtree.
+    ///
+    /// Note: it doesn't return whether the key exists in the subtree, so caller
+    /// need to ensure the logic.
+    #[allow(dead_code)]
+    pub(crate) fn delete(mut self: Box<Self>, key: &Range) -> Option<Box<Self>> {
+        match self.key.cmp(key) {
+            Ordering::Equal => {
+                return self.delete_root();
+            }
+            Ordering::Less => {
+                if let Some(node) = self.right.take() {
+                    let right = node.delete(key);
+                    self.right = right;
+                    return Some(self.update_height_and_balance());
```

Comment:
nit: functions named with `and` are typically a yellow flag because it means we're not following the single responsibility principle at the level of that function. In this case, it would be better to just call `update_height` and `rebalance` even if there are two calls.